### PR TITLE
Add gradient buffers to CollaborativeOptimizer

### PR DESCRIPTION
**Pull request by justheuristic**
- added an option to accumulate gradients locally (default = True, can be disabled to save some memory, not critical in tutorials)
- fixed a bug in CollaborativeOptimizer that caused it to implicitly scale all updates by self.local_steps_accumulated
- fixed typos in description

### CODE REVIEW

File: hivemind/client/optim/collaborative.py [hunk 1]
```diff
@@ -69,6 +68,12 @@ class CollaborativeOptimizer(DecentralizedOptimizerBase):
     :param metadata_expiration: peer's metadata (e.g. samples processed) is stored onto DHT for this many seconds
     :param averaging_timeout: if an averaging step hangs for this long, it will be cancelled.
     :param scheduler: if specified, use this scheduler to update optimizer learning rate
+    :param reuse_grads_accumulators: if True, use model's .grad buffers for gradient accumulation.
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:71]:
```suggestion
    :param reuse_grad_accumulators: if True, use model's .grad buffers for gradient accumulation.
```

File: hivemind/client/optim/collaborative.py [hunk 2]
```diff
@@ -69,6 +68,12 @@ class CollaborativeOptimizer(DecentralizedOptimizerBase):
...
+      This is more memory efficient, but it requires that the user does *NOT* call model/opt zero_grad at all
+    :param accumulate_grads_on: if specified, accumulate gradients on this device. By default, this will use the same
+     device as model parameters. One can specify a different device (e.g. 'cpu' vs 'cuda') to save device memory at
+     the cost of extra time per step. If reuse_gradient_accumulators is True, this parameter has no effect.
+    :param kwargs: additional parameters are forwarded to DecentralizedAverager parameters
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:76]:
```suggestion
    :param kwargs: additional parameters forwarded to DecentralizedAverager
```

File: hivemind/client/optim/collaborative.py [hunk 3]
```diff
@@ -134,9 +142,11 @@ def step(self, batch_size: Optional[int] = None, **kwargs):
         :param batch_size: optional override for batch_size_per_step from init
         :note: this .step is different from normal pytorch optimizers in several key ways. See __init__ for details.
         """
-        if batch_size is not None and self.batch_size_per_step is None:
-            raise ValueError("Please either set batch_size_per_step parameter at init or provide batch_size in .step")
-        batch_size = self.batch_size_per_step if batch_size is None else batch_size
+        if self.batch_size_per_step is None:
+            if batch_size is None:
+                raise ValueError("Please either set batch_size_per_step parameter at init or when calling .step")
+            self.batch_size_per_step = batch_size
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:148]: We should explicitly log that we're overriding the field with the passed value

File: hivemind/client/optim/collaborative.py [hunk 4]
```diff
@@ -184,6 +199,43 @@ def step(self, batch_size: Optional[int] = None, **kwargs):
             logger.log(self.status_loglevel, f"Optimizer step: done!")
             return output
 
+    @property
+    def _grad_buffers(self) -> Sequence[torch.Tensor]:
+        """ pytorch-internal gradient buffers """
+        return [param.grad if param.grad is not None else torch.zeros_like(param)
```
Comment (mryab) [inline hivemind/client/optim/collaborative.py:205]: Rewrite to generator?

**Review by mryab — CHANGES_REQUESTED** [review]

**Review by leshanbog — APPROVED** [review]

**Review by mryab — APPROVED** [review]
